tv/tv.cpp

```diff
@@ -616,32 +649,40 @@ llvmGetPassPluginInfo() {
             return;
 
           // Run only when it is at the boundary
-          bool is_first = TVPass::pass_name.empty();
-          bool do_start = !TVPass::batch_started && do_skip(TVPass::pass_name)
+          bool is_first = pass_name.empty();
+          bool do_start = !TVPass::batch_started && do_skip(pass_name)
               && !do_skip(P);
-          bool do_finish = TVPass::batch_started && !do_skip(TVPass::pass_name)
+          bool do_finish = TVPass::batch_started && !do_skip(pass_name)
               && do_skip(P);
 
           if (do_start)
-            TVPass::batched_pass_begin_name = TVPass::pass_name;
+            TVPass::batched_pass_begin_name = pass_name;
           else if (is_first)
             TVPass::batched_pass_begin_name = "beginning";
 
+          if ((is_first || do_start) && opt_save_ir && !report_filename.empty())
```

Comment:
I think this bit is not needed.
For Alive2 verification we have to do this because the callback is after each pass, but since you are running a before callback, it shouldn't be needed. (running the before callback is the right choice for cloning the module)

Reply:
ok let's talk through this a bit...

by "this bit" do you mean the `(is_first || do_start)` check? if so, I think these are needed because we want to save the IR at the start of the entire batch of passes, instead of saving it before each pass in the batch. 

there's probably a bit more work to do here where we record the names of all passes in the batch so we can replay them in that order, to repro any bugs that get found, but I don't want to deal with that right now.

or perhaps we don't need to save IR at all for the batched case, since a bug found with batching will generally want to be reproduced with batching turned off?

Reply:
Oh, sorry, I didn't realize this was in the big if for batched ops.
Ignore my comment then.

Can I merge this PR?

Reply:
yes please merge